common/vendor/pexpect.py

```diff
@@ -0,0 +1,2134 @@
+
+        # If command is an int type then it may represent a file descriptor.
+        if isinstance(command, type(0)):
+            raise ExceptionPexpect('Command is an int type. ' +
+                    'If this is a file descriptor then maybe you want to ' +
+                    'use fdpexpect.fdspawn which takes an existing ' +
+                    'file descriptor instead of a command string.')
+
+        if not isinstance(args, type([])):
+            raise TypeError('The argument, args, must be a list.')
+
+        if args == []:
+            self.args = split_command_line(command)
+            self.command = self.args[0]
+        else:
+            # Make a shallow copy of the args list.
+            self.args = args[:]
+            self.args.insert(0, command)
+            self.command = command
+
+        command_with_path = which(self.command)
+        if command_with_path is None:
+            raise ExceptionPexpect('The command was not found or was not ' +
+                    'executable: %s.' % self.command)
+        self.command = command_with_path
+        self.args[0] = self.command
+
+        self.name = '<' + ' '.join(self.args) + '>'
+
+        assert self.pid is None, 'The pid member must be None.'
+        assert self.command is not None, 'The command member must not be None.'
+
+        if self.use_native_pty_fork:
+            try:
+                self.pid, self.child_fd = pty.fork()
+            except OSError:  # pragma: no cover
+                err = sys.exc_info()[1]
+                raise ExceptionPexpect('pty.fork() failed: ' + str(err))
+        else:
+            # Use internal __fork_pty
+            self.pid, self.child_fd = self.__fork_pty()
+
+        # Some platforms must call setwinsize() and setecho() from the
+        # child process, and others from the master process. We do both,
+        # allowing IOError for either.
+
+        if self.pid == pty.CHILD:
+            # Child
+            self.child_fd = self.STDIN_FILENO
+
+            # set default window size of 24 rows by 80 columns
+            try:
+                self.setwinsize(24, 80)
+            except IOError as err:
+                if err.args[0] not in (errno.EINVAL, errno.ENOTTY):
+                    raise
+
+            # disable echo if spawn argument echo was unset
+            if not self.echo:
+                try:
+                    self.setecho(self.echo)
+                except (IOError, termios.error) as err:
+                    if err.args[0] not in (errno.EINVAL, errno.ENOTTY):
+                        raise
+
+            # gwenzek, HACK: I removed this line because I wanted to get rid of 'rresource' dependency,
+            # but I have no idea what it was supposed to do.
+            #
+            # Do not allow child to inherit open file descriptors from parent.
+            # max_fd = resource.getrlimit(resource.RLIMIT_NOFILE)[0]
+            # os.closerange(3, max_fd)
+            os.closerange(3, 1024)
```

Comment:
This part is a bit hacky, I don't have any knowledge of how "os.closerange" works. On my computer `resource.getrlimit(resource.RLIMIT_NOFILE)` returns `(1024, 65536)` that's why I hardcoded 1024 here